Coding guidelines (#181)
GridTools/gt4py: CONTRIBUTING.rst

Review Process
--------------

Very large pull requests may receive only a partial review at first.
This is to be able to respond within a reasonably short time frame while
still taking time to fairly assess all the changes.

We use CODING_GUIDELINES_ as a reference for reviewing contributions from
a code style and design standpoint.

Further points to be considered are:

- Check all authors are covered by a contributor agreement and added to AUTHORS_

- All added or changed functionality must have a test that fails before and
  passes after changes and documents the correct behaviour.

- The documentation (docstrings as well as ``/docs`` must be kept in sync.

- New files are not allowed to skip ``pre-commit`` checks. Look for additions
  to exclude lists in  pre-commit-config_. The only exception is that
  test modules might skip ``mypy`` checks.

- If a file currently excluded from checks in pre-commit-config_  is changed
  substantially, the reviewer may request dropping it from the exclude list and
  cleaning up the remaining issues.  The remaining issues must be small
  compared to the size of the PR.

> - Any comments about future work required (``#todo`` and other fixme-style
>   comments) should be either addressed before merging or deleted after an issue
>   is opened describing the future changes.

Reviewer: ```suggestion
```
We discussed this topic, ok to remove? Then let's merge.

Reviewer 2: I agree. Let's remove this paragraph and merge the PR (or close it, it's been open for too long). @DropD Do you also agree?

Contributing
============

Contributions to GT4Py are welcome, and they are greatly appreciated, whether
they be a well-described issue or contributions to the GT4Py code.

Proper credit will be given to contributors by adding their names to the
AUTHORS_ file. ETH Zurich owns the GT4Py library, and external
contributors must sign a contributor assignment agreement.

Bug reports and feature requests
--------------------------------

We always welcome both, don't hesitate opening an issue directly in the `GitHub
repository <https://github.com/GridTools/gt4py/issues>`_. If possible please
add a minimal snippet (or jupyter notebook) of code that illustrates the
problem or feature. The clearer the description the quicker we can get back to
you!

Contributing code
-----------------

> We use a fork and pull request workflow for code contributions. If that does
> not sound familiar to you, do not worry! A simple search for ``github fork pull
> request workflow`` will yield many explanations and tutorials.
> 

Reviewer 2: ```suggestion
We use a [fork and pull request](https://www.atlassian.com/git/tutorials/comparing-workflows/forking-workflow) workflow for code contributions.
```

> Pull requests need to pass some automated checks as well as a review before
> they can be merged. If you set up your development environment as recommended
> for developers in the README_, it is easy to run the automated checks
> locally before creating your pull request.

Reviewer 2: ```suggestion
Pull requests need to pass all the automated checks as well as a review before
they can be merged. You should set up your development environment as
recommended for developers in the README_, and run the automated checks
locally before creating your pull request.
```
